Feature/ha discovery

dominikandreas opened the pull request:
This is a work in progress, but wanted to get some early feedback. The main contributions of this PR are:

- Implemented support for home assistant discovery (https://github.com/DennisOSRM/hms-mqtt-publisher/issues/17) 
- Add extra sensors (https://github.com/DennisOSRM/hms-mqtt-publisher/issues/14) 

Other notable changes / additions:
- added vscode devcontainer support and debug launch configuration to simplify development for future contributors
- add serde dependency for json serialization
- version bump to 0.2.0 because of breaking changes (mqtt paths change)
- refactor log init into its own module
- add mock_state function to inverter for testing purposes

Note that this is my first time developing in rust and there's probably a lot of things to improve / things that an experienced rust developer would've probably done differently. I'm happy to receive some constructive feedback and improve where necessary :-)

TODO:
- update home assistant addon once PR is merged / docker build is ready
- make home assistant prefix configurable (https://github.com/DennisOSRM/hms-mqtt-publisher/issues/25)

DennisOSRM commented:
Thanks for the PR. It will take a couple of days before I can get around to reviewing this. 

From a quick glance, it seems this PR contains a couple of things that likely should be seperate PRs, and the testing code perhaps should not bleed into the implementation. That being said, this isn't a comprehensive review and just an initial impression. 

Thanks for the contribution, please keep up the great work.

DennisOSRM commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1 @@
+FROM rust:buster
```
The files in .devcontainer should not be part of this PR.

DennisOSRM commented on .vscode/launch.json:
```diff
@@ -0,0 +1,21 @@
+{
```
The files in `.vscode/*` should not be part of this PR.

DennisOSRM commented on src/inverter.rs:
```diff
@@ -39,7 +40,42 @@ impl<'a> Inverter<'a> {
     }
 
-    pub fn update_state(&mut self) -> Option<HMSStateResponse> {
+    pub fn mock_state(&mut self) -> Option<HMSStateResponse> {
```
The test code should be separated from the implementation. This could be done in a tool simulating the behavior of an actual inverter. Issue #11 is tracking such a feature. This should be dropped from the PR.

DennisOSRM commented on src/main.rs:
```diff
@@ -62,11 +52,15 @@ fn main() {
 
         // TODO: this has to move into the Inverter struct in an async implementation
-        thread::sleep(Duration::from_millis(REQUEST_DELAY));
+        if !cli.test{
```
See above for the comment that API simulation should implemented in #11.

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+impl HMSStateResponse {
+    fn get_model(&self) -> String {
+         // TODO: identify model from dtu_sn
+        "HMS-800W-T2".to_string()
```
This is not necessarily the inverter. While tested, the tool likely works with all devices of the HMS-XXXX-2T series.

dominikandreas replied:
Right, since we currently don't have a way to determine the exact model name, I've removed the mention of 800W-2T

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+        ).unwrap_or(&InverterState::default()).clone()
+    }
+
+    fn to_json_payload(&self) -> serde_json::Value {
```
consider using the [protobuf-json](https://crates.io/crates/protobuf-json) crate, which makes this function a one-liner basically.

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+/// `DeviceConfig` is used to define the configuration for a Home Assistant device
+/// in the MQTT discovery protocol and is used to group entities together.
+/// 
+#[derive(Serialize, Clone)]
```
consider moving the struct into a separate source file

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+}
+
+
+impl DeviceConfig {
```
consider moving the struct into a separate source file

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+            identifiers,
+            manufacturer: "Hoymiles".to_string(),
+            /// Rust compiler sets the CARGO_PKG_VERSION environment from the Cargo.toml . 
+            sw_version: env!("CARGO_PKG_VERSION").to_string(),
```
This is arguably not part the version of the inverter device

dominikandreas replied:
No, but I thought it makes sense to log the version of the publisher here (we can also change it in the future in case we figure out a way to extract the sw version of the inverter)

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+/// https://developers.home-assistant.io/docs/core/entity/sensor/
+///
+#[derive(Serialize)]
+pub struct SensorConfig {
```
consider moving this to a separate source file

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+}
+
+
+fn create_sensor_configs(hms_state: &HMSStateResponse, state_topic: &str) -> Vec<SensorConfig> {
```
should this be a `::from(...)` member function rather than a free factory?

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -42,35 +251,41 @@ impl Mqtt {
-        self.client
-            .subscribe("hms800wt2/pv_current_power", QoS::AtMostOnce)
-            .unwrap();
-        match self.client.publish(
```
this seems like a breaking change.

DennisOSRM reviewed (changes requested):
Thanks again for filing this PR. I think it contains very interesting and worthwhile ideas. Since the PR is rather big and contains several unrelated changes, I'd suggest to do the following:


- [ ] remove the development related files and directories
- [ ] the testing code should be done in a distinct PR without having to branch between test and non-test code. I.e. the implementation should be oblivious to the fact whether it is run in a test environment or not.
- [ ] move structs into their own source file
- [ ] consider using `protobuf_json` crate instead of the rolling your own conversion to json
- [ ] implement the auto discovery in a non-breaking way

While the above may seem like a lot, I think the PR is a great first stab into the right direction. The intent for the PR is great, please keep up the good work.

dominikandreas commented:
Thanks for your comprehensive review!

> remove the development related files and directories

I will create a separate PR for this

> the testing code should be done in a distinct PR without having to branch between test and non-test code. 

Good point. Didn't have time to address this properly so just hacked in something quickly that allowed me to do some minimal testing during the addon development. I will remove the testing related code for now

> move structs into their own source file

Will do. Wasn't sure how to structure it or whether it's necessary to split the mqtt implementation up in two parts, one for home assistant, one without home assistant support.

> consider using protobuf_json crate instead of the rolling your own conversion to json

Thought about this as well, but since the values are scaled during the serialization, I decided to just do it the manual way. I assume the protobuf schemas won't often if at all, so I hope that's okay. 

>  implement the auto discovery in a non-breaking way

The breaking change is unfortunate. But apart from home assistant support, another upside is that it would allow utilizing multiple inverters in the future (part of the serial number is added to the topic). Another part of the breaking change is that the whole update is sent json-encoded to a single topic instead of splitting it to multiple topics. This should make it more efficient in combination with home assistant (and possibly other endpoints)

DennisOSRM commented:
> The breaking change is unfortunate. But apart from home assistant support, another upside is that it would allow utilizing multiple inverters in the future (part of the serial number is added to the topic). Another part of the breaking change is that the whole update is sent json-encoded to a single topic instead of splitting it to multiple topics. This should make it more efficient in combination with home assistant (and possibly other endpoints)

That's a good point. Thanks for the clarification. I think the json-encoded single topic could be done independently, is that right? If so, we could first add things, and consider the breaking change at a later point. Would this make sense?

dominikandreas commented:
Yes, we could change the home assistant device configurations and publish on separate topics to delay the breaking change. But if we're certain it's going to happen sooner or later (and I think supporting multiple inverters is a enough good reason), delaying the breaking change will only cause more users to be affected. 

While we're at it we should also remove the "800W-2T" references, e.g. from the topic, to prevent a future breaking change.

DennisOSRM commented:
Ok, that's a convincing argument. Let's do the breaking change now, and aim at the same time to minimize the size of this PR.

dominikandreas commented:
I think I addressed all of the review comments and fixed some additional bugs. Here's an example of the provided data in home assistant:

<img width="604" alt="image" src="https://github.com/DennisOSRM/hms-mqtt-publisher/assets/13525040/bcee56b4-9d3c-428e-9a56-4550f8f33cfc">

In order to enable these changes in the addon, we first need an image on docker hub that I can reference in the Dockerfile of the home assistant addon. I have a build running right now, but it's taking a while. I will hopefully push another update tomorrow when it's done

DennisOSRM commented:
Very cool. Please have a bit of patience, since it will again take a couple of days before I can get to it.

DennisOSRM commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "hms-mqtt-publish"
-version = "0.1.0"
+version = "0.2.0"
```
minor nitpick. let's leave this at 0.1.0 for now. The version will be increased as part of the release process.

DennisOSRM commented on .gitignore:
```diff
@@ -14,4 +14,4 @@ Cargo.lock
 # Ignore VS Code user settings
-.vscode/
\ No newline at end of file
+.vscode/
```
please drop this white-space change

DennisOSRM commented on src/inverter.rs:
```diff
@@ -38,7 +38,7 @@ impl<'a> Inverter<'a> {
         }
     }
-
+    
```
please run `cargo fmt` to ensure proper formatting and removal of stray whitespace

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -42,35 +47,140 @@ impl Mqtt {
+    fn publish_json(&mut self, topic: &str, payload: serde_json::Value) {
+        debug!("Publishing to {topic} with payload {payload}");
+
+        let payload = serde_json::to_string(&payload).unwrap();
```
why not use the `protobuf_json` crate to publish the original response?

dominikandreas replied:
did you mean regarding line 77? The `hms_state.to_json_payload` includes the scaling and the additional efficiency attribute. wasn't sure how to do that without manually implementing the json conversion

DennisOSRM replied:
Makes sense.

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -42,35 +47,140 @@ impl Mqtt {
+        }
+    }
+
+    fn to_json_payload(&self) -> serde_json::Value {
```
one thing that I haven't fully understood, yet of this is a json schema that HA requires. Is there any formal requirement, and where can I find some information on this?

dominikandreas replied:
Home assistant requries two parts: config messages and state messages. The config messages define the sensors and corresponding state topics to listen for state messages. 

The state messages don't require any specific schema. Rather the configuration messages contain a "value_template" (`SensorConfig.value_template`), which is used to extract the state value from a json-encoded message sent to the "state_topic" (`SensorConfig.state_topic`). The template language used is jinja.

We've defined the template for each sensor as `let value_template = format!("{{{{ value_json.{} }}}}", unique_id);`, whereas `value_json` would correspond to the json-serialized `HMSStateResponse` and `unique_id` is defined to match a corresponding attribute in `HMSStateResponse`

DennisOSRM commented on src/mqtt_schemas.rs:
```diff
@@ -0,0 +1,114 @@
+
```
the file should be named after the struct in it, or the most important one.

dominikandreas replied:
yes, wasn't happy with the naming either, but couldn't think of anything else. Feel free to suggest an alternative or directly rename it to something else

DennisOSRM commented on src/protos/RealData.proto:
```diff
@@ -12,7 +12,7 @@ message InverterState {
   int32 port_id = 2;
   int32 grid_voltage = 3;     // [V], factor 0.1
-  int32 grid_freq = 4;        // [Hz], factor 0.1
+  int32 grid_freq = 4;        // [Hz], factor 0.01
```
arguably, the protobuf change should go in independently of this PR.

DennisOSRM commented on src/main.rs:
```diff
@@ -31,19 +31,7 @@ struct Cli {
 static REQUEST_DELAY: u64 = 30_500;
 
 fn main() {
-    Builder::new()
```
can we move the non-functional change into it's own PR?

DennisOSRM reviewed (commented):
- [ ] run `cargo fmt`
- [ ] move non-functional changes into distinct PRs to separate concerns

This is looking really close. Couple of remarks on the PR that shouldn't take too long to resolve. Then we can get this in.

To give some perspective of where I am taking the architecture of the tool are probably helpful. Similar to the idea of supporting several inputs, I'd like the tool to support several output channels (for a lack of a better name). I.e. there's a simple MQTT publisher, there's the HA autodiscovery and publishing of this PR, and other outputs like InfluxDB and perhaps others. The user will be able to configure where to publish.

Once this PR is ready, I want to start implementing these channels by building right on top of this branch and then merge the two in short succession once both are ready.

DennisOSRM commented:
I did a very hasty implementation of output channels (cf. https://github.com/DennisOSRM/hms-mqtt-publisher/pull/27#pullrequestreview-1746689155) on top of your PR: https://github.com/DennisOSRM/hms-mqtt-publisher/pull/34/commits/7460c0e6194d36540d5682457d884d2e632e5074

DennisOSRM commented on src/mqtt.rs:
```diff
@@ -1,9 +1,12 @@
 use std::{thread, time::Duration};
```
in light of the _output channel_ feature (see below) it probably makes sense to rename this file to `home_assistent.rs` since it will implement the output channel to a HA instance.

DennisOSRM reviewed (approved):
Will merge changes and do a couple of minor changes on top. Thanks so much for the contribution

dominikandreas commented:
Sorry I was preoccupied with my day job and family. I'll have a look at the output channels PR

DennisOSRM commented:
No worries. Life comes first, hacking on code second. 😉
